### PR TITLE
Nullable arrays in models

### PR DESCRIPTION
**Pull request by ondojo**
Fix array notation bug related to #5040

_Issue #, if available:_ #5040 

_Description of changes:_ Validation of model ignored the isRequired attribute in case the attribute value was of type array. Added this consideration to the if clause to allow for graphlQL annotations such as `[foobar]` or `[foobar!]`. Now the attributes can be set to null or empty arrays and the sync still works.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by ondojo** [comment]
Hello,

the circle ci error message says 

```bash
Creating workspace archive...

Uploading workspace archive...

Error uploading workspace archive: AccessDenied: Access Denied
status code: 403, request id: 444FF68393D2C9DC, host id: 5giQT2G09PD0Ig+4Ei+SZLZXJo1sYyleelJIH9wZB15H9ZAAcjO5WKcN0Q6Yeb456o/6lO2INlg=
```

Is there any Chance we can fix this?

**Comment by mauerbac** [comment]
hi @ondojo - sorry for the delay. I just looked at the CircleCi tests on my end and it looks like one of the datastore unit tests failed. Are you able to run these locally and update the test? Thanks for the contribution

File: packages/datastore/src/datastore/datastore.ts [hunk 1]
```diff
@@ -241,13 +241,13 @@ const validateModelFields = (modelDefinition: SchemaModel | SchemaNonModel) => (
...
 			if (isArray) {
-				if (!Array.isArray(v)) {
+				if (!Array.isArray(v) && isRequired) {
 					throw new Error(
 						`Field ${name} should be of type ${jsType}[], ${typeof v} received. ${v}`
 					);
 				}
 
-				if ((<[]>v).some(e => typeof e !== jsType)) {
+				if (v !== null && (<[]>v).some(e => typeof e !== jsType)) {
```
Comment (seang96) [inline packages/datastore/src/datastore/datastore.ts:250]: I believe this should check for v being undefined as well. I have two reasons for this:

1. The require check at line 236 checks for undefined as well
2. Typescript allows copyOf records to be undefined but not null

Another note, 2 opens up a new issue. Setting the value to undefined will result in it being ignored as an input in the graphql request resulting in the field not being updated. If set to null it works, but Typescript checks says null is not valid in the copyOf mutation. I am unsure if this is due to this PR or it is an existing issue with DataStore that I discovered upon testing this PR.

**Comment by sammartinez** [comment]
Resolving as the above PR looks to have been taken care of in a separate pull request which has be related to this pull request.

**Pull request closed without merging**